src/dag/lazy-store.ts

```diff
@@ -0,0 +1,551 @@
+    if (oldCacheEntry) {
+      this._cacheEntries.delete(hash);
+      this._cacheEntries.set(hash, oldCacheEntry);
+      return;
+    }
+
+    // Only cache if there is a ref from a head to this chunk
+    const refCount = this._refCounts.get(hash);
+    if (refCount === undefined || refCount < 1) {
+      return;
+    }
+    const valueSize = this._getSizeOfValue(chunk.data);
+    if (valueSize > this._cacheSizeLimit) {
+      // This value cannot be cached due to its size exceeding the
+      // cache size limit, don't evict other entries to try to make
+      // room for it.
+      return;
+    }
+    this._size += valueSize;
+    this._cacheEntries.set(hash, {chunk, size: valueSize});
+    chunk.meta.forEach(refHash => {
+      this._refCounts.set(refHash, (this._refCounts.get(refHash) || 0) + 1);
+    });
+
+    this._ensureCacheSizeLimit();
+  }
+
+  private _ensureCacheSizeLimit() {
+    if (this._size <= this._cacheSizeLimit) {
+      return;
+    }
+    for (const entry of this._cacheEntries) {
+      if (this._size <= this._cacheSizeLimit) {
+        break;
+      }
+      this.delete(entry[1]);
+    }
+  }
+
+  delete(cacheEntry: CacheEntry): void {
+    const {hash} = cacheEntry.chunk;
+    this._size -= cacheEntry.size;
+    this._cacheEntries.delete(hash);
+    cacheEntry.chunk.meta.forEach(refHash => {
+      const oldCount = this._refCounts.get(refHash);
+      assertNotUndefined(oldCount);
+      assert(oldCount > 0);
+      const newCount = oldCount - 1;
+      if (newCount === 0) {
+        this._refCounts.delete(refHash);
+        const refCacheEntry = this._cacheEntries.get(refHash);
+        if (refCacheEntry) {
+          this.delete(refCacheEntry);
+        }
+      } else {
+        this._refCounts.set(refHash, newCount);
+      }
+    });
+  }
+
+  updateForCommit(
+    chunksToPut: Iterable<Chunk>,
+    hashesToDelete: Iterable<Hash>,
+  ): void {
+    // Commit has already updated this._refCounts to reflect these puts and
+    // deletes.  First we do all the puts and deletes to bring this._refCounts
+    // and this._cacheEntries into a consistent state. Then we ensure that the
+    // cache is below its size limit, using this.delete and
+    // this._ensureCacheSizeLimit, both of which require this._refCounts and
+    // this._cacheEntries to be in a consistent state to work correctly.
+    const cacheEntiresLargerThanLimit = [];
+    for (const chunk of chunksToPut) {
+      const {hash} = chunk;
+      const oldCacheEntry = this._cacheEntries.get(hash);
+      if (oldCacheEntry) {
+        // If there is an existing cache entry then the cached value must be
+        // equivalent.  Update order in map for LRU tracking but avoid
+        // recomputing size and creating a new cache entry.
+        this._cacheEntries.delete(hash);
+        this._cacheEntries.set(hash, oldCacheEntry);
+      } else {
+        const valueSize = this._getSizeOfValue(chunk.data);
+        this._size += valueSize;
```

Comment:
I'm not sure this is correct? If the value is larger than the limit I don't think you should update the size here?

Reply:
OK. Makes sense